src/server/accessors/ExternalComponentsExtend.ts

```diff
@@ -0,0 +1,12 @@
+import { AppExternalComponentManager } from '../managers/AppExternalComponentManager';
+
+import { IExternalComponentsExtend } from '../../definition/accessors';
+import { IExternalComponent } from '../../definition/externalComponent/IExternalComponent';
+
+export class ExternalComponentsExtend implements IExternalComponentsExtend {
+    constructor(private readonly manager: AppExternalComponentManager, private readonly appId: string) { }
+
+    public async register(externalComponent: IExternalComponent): Promise<void> {
```

Comment:
What will happen when an App calls register multiple times with different `IExternalComponent` instances? Based upon how the precedence of the slash commands, it _should_ add the different instances and have all of them being valid. However, looking elsewhere in this external component code it does not behave the same way. We need the behavior/logic to be consistent if we are using the same terminology within the codebase.

Reply:
Still wondering what happens when they call this method twice and provide different components. Will this error out or will it fail silently or will it work and if it works then which instance is valid?

Reply:
Hi @graywolf336 . The `register` method actually calls the method `addExternalComponnet` of the AppExternalComponent. 
- when we call this method twice and provide different components, the app will own 2 external components.
- when we call this method twice and provide the same components, the will still be an external component (we store external components in a map).
- when we call this method twice and provide the different components **but with the same name**, we will throw an ExternalComponentAlreadyTouchedError.
https://github.com/RocketChat/Rocket.Chat.Apps-engine/blob/adc01ab56f7cbd945ce395de24aa56b172fb6685/src/server/managers/AppExternalComponentManager.ts#L74-L80

---

Below is the latest implementation and unit tests:

https://github.com/RocketChat/Rocket.Chat.Apps-engine/blob/83d2ba18c859fe3eaf87d634ff1c2e3b640beb63/src/server/managers/AppExternalComponentManager.ts#L57-L81

https://github.com/RocketChat/Rocket.Chat.Apps-engine/blob/83d2ba18c859fe3eaf87d634ff1c2e3b640beb63/tests/server/managers/AppExternalComponentManager.spec.ts#L85-L110